File: cps.py
import numpy as np
import argparse

N = 100


def inc(op):
    def inc_inner(k):
        op(lambda x: k(x + 1))

    return inc_inner


def dbl(op):
    def dbl_inner(k):
        op(lambda x: k(x * 2))

    return dbl_inner


def prt(x):
    ##print(f"{x}, ", end="")
    pass


def scan(xs: list[int]):
    # currying
    # closed over xs
    # is basically op(k) now
    def scn_inner(k):
        for x in xs:
            k(x)

    return scn_inner


def _parse_args():
    parser = argparse.ArgumentParser(prog="cps.py", description="CPS implementation")

    parser.add_argument("-n", default=N, type=int, help="array size")
    parser.add_argument("-d", default=0, type=int, help="default value for the array")

    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = _parse_args()
    xs = [int(args.d)] * args.n

    dbl(inc(dbl(inc(scan(xs)))))(prt)
    print("")

File: cps.cpp
#include "args.h"

#include <iostream>
#include <functional>
#include <vector>

using namespace std;

#define N 100

template <class T>
auto inc(T op)
{
    return [op](auto k)
    { op([k](int x)
         { k(x + 1); }); };
}

template <class T>
auto dbl(T op)
{
    return [op](auto k)
    { op([k](int x)
         { k(x * 2); }); };
}

void prt(int x)
{
    // cout << x << ", ";
}

auto scan(const vector<int> &xs)
{
    return [xs](auto k)
    {
        for (int x : xs)
        {
            k(x);
        }
    };
}

int main(int argc, char const *argv[])
{
    /* argument parsing */
    long n = N;
    int default_value = 0;
    if (parse_args(argc, argv, &n, &default_value) != 0)
        return 1;

    cout << "n:" << n << endl;

    /* array xs */
    vector<int> xs(n, default_value);

    // scan(xs)(prt);
    // inc(scan(xs))(prt);
    dbl(inc(dbl(inc(scan(xs)))))(prt);
    cout << endl;

    return 0;
}
